### ISSUE TITLE
Missing warning for an unused const in async fn

### ISSUE THREAD

**Issue by stepantubanov** [rust-lang/rust#108296]
I tried this code:

```rust
pub async fn foo() {
    const A: usize = 1;
}
```

I expected to see the warning:
> warning: constant `A` is never used

Instead no warning was reported by the compiler.

Observed in nightly, 1.67.
Seems like the warning was shown in 1.63 and stopped showing in 1.64.

**Comment by clubby789** [comment]
Bisected to
```
  commit[0] 2022-06-30UTC: Auto merge of 98377 - davidv1992:add-lifetimes-to-argument-temporaries, r=oli-obk
  commit[1] 2022-07-01UTC: Auto merge of 98706 - flip1995:clippyup, r=Dylan-DPC
  commit[2] 2022-07-01UTC: Auto merge of 98752 - matthiaskrgr:rollup-uwimznc, r=matthiaskrgr
  commit[3] 2022-07-01UTC: Auto merge of 98730 - matthiaskrgr:rollup-2c4d4x5, r=matthiaskrgr
  commit[4] 2022-07-01UTC: Auto merge of 98402 - cjgillot:undead, r=michaelwoerister
  commit[5] 2022-07-01UTC: Auto merge of 98767 - Dylan-DPC:rollup-j1gq5sr, r=Dylan-DPC
  commit[6] 2022-07-01UTC: Auto merge of 93967 - cjgillot:short-struct-span, r=petrochenkov
```
CI has expired so I couldn't automatically bisect further

**Comment by clubby789** [comment]
It looks like the issue is that dead code detection doesn't work inside closures - 
```rs
#[deny(dead_code)]
#[allow(unused_must_use)]
pub fn foo() {
    || {
        fn a() {}
    };
}
```
This is denied by 1.63 but allowed on 1.64+.
@rustbot label +regression-from-stable-to-stable

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23108296.20Missing.20warning.20for.20an.20unused.20const.20in.20async.20fn)).

@rustbot label -I-prioritize +P-medium
